spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskService.java

```diff
@@ -171,75 +179,110 @@ public DefaultTaskService(DataSourceProperties dataSourceProperties,
-		Assert.notNull(taskDeploymentProperties, "The provided runtimeProperties must not be null.");
-
-		if (maxConcurrentExecutionsReached()) {
-			throw new IllegalStateException(String.format(
-					"The maximum concurrent task executions [%d] is at its limit.",
-					taskConfigurationProperties.getMaximumConcurrentTasks()));
-		}
+		RequestCriteria requestCriteria = createLaunchRequest(taskName, taskDeploymentProperties,commandLineArgs);
 
-		TaskDefinition taskDefinition = this.taskDefinitionRepository.findOne(taskName);
-		if (taskDefinition == null) {
-			throw new NoSuchTaskDefinitionException(taskName);
-		}
-		TaskParser taskParser = new TaskParser(taskDefinition.getName(), taskDefinition.getDslText(), true, true);
-		TaskNode taskNode = taskParser.parse();
-		// if composed task definition replace definition with one composed task
-		// runner and executable graph.
-		if (taskNode.isComposed()) {
-			taskDefinition = new TaskDefinition(taskDefinition.getName(),
-					TaskServiceUtils.createComposedTaskDefinition(
-							taskNode.toExecutableDSL(), taskConfigurationProperties));
-			taskDeploymentProperties = TaskServiceUtils.establishComposedTaskProperties(taskDeploymentProperties,
-					taskNode);
-		}
-
-		AppRegistration appRegistration = this.registry.find(taskDefinition.getRegisteredAppName(),
-				ApplicationType.task);
-		Assert.notNull(appRegistration, "Unknown task app: " + taskDefinition.getRegisteredAppName());
-		Resource appResource = this.registry.getAppResource(appRegistration);
-		Resource metadataResource = this.registry.getAppMetadataResource(appRegistration);
-
-		TaskExecution taskExecution = taskExecutionRepository.createTaskExecution(taskName);
-		taskDefinition = TaskServiceUtils.updateTaskProperties(taskDefinition,
-				dataSourceProperties);
-
-		Map<String, String> appDeploymentProperties = new HashMap<>(commonApplicationProperties.getTask());
-		appDeploymentProperties.putAll(
-				TaskServiceUtils.extractAppProperties(taskDefinition.getRegisteredAppName(), taskDeploymentProperties));
-
-		Map<String, String> deployerDeploymentProperties = DeploymentPropertiesUtils
-				.extractAndQualifyDeployerProperties(taskDeploymentProperties, taskDefinition.getRegisteredAppName());
-		if (StringUtils.hasText(this.dataflowServerUri) && taskNode.isComposed()) {
-			TaskServiceUtils.updateDataFlowUriIfNeeded(this.dataflowServerUri, appDeploymentProperties,
-					commandLineArgs);
-		}
-		AppDefinition revisedDefinition = TaskServiceUtils.mergeAndExpandAppProperties(taskDefinition, metadataResource,
-				appDeploymentProperties, this.whitelistProperties);
-		List<String> updatedCmdLineArgs = this.updateCommandLineArgs(commandLineArgs, taskExecution);
-		AppDeploymentRequest request = new AppDeploymentRequest(revisedDefinition, appResource,
-				deployerDeploymentProperties, updatedCmdLineArgs);
-		String id = this.taskLauncher.launch(request);
+		String id = this.taskLauncher.launch(requestCriteria.getAppDeploymentRequest());
```

Comment:
What if the taskLauncher throws an exception, then there has been a commit into the database for something that hasn't launched.

Reply:
@markpollack The TaskExecution table will have an entry with the Task_Execution_id and nothing else.  It would be similar to the case where a Task failed to launch (for whatever reason).   Except in this case there would be a stack trace in SCDF.